File: лр 3/.py
 #**БИНАРНАЯ КУЧА**
import heapq

max_heap = []
heapq.heappush(max_heap, -10)  # храним -10
heapq.heappush(max_heap, -5)   # храним -5
heapq.heappush(max_heap, -20)  # храним -20

# Извлекаем максимум (берём минус от минимального в куче)
max_val = -heapq.heappop(max_heap)
print(max_val)  # 20


 #**БИНОМИНАЛЬНАЯ КУЧА**
class BinomialHeapNode:
    def __init__(self, key):
        self.key = key
        self.children = []  # список дочерних узлов
        self.parent = None
        self.degree = 0  # ранг (число детей)

class BinomialHeap:
    def __init__(self):
        self.roots = []  # список корней биномиальных деревьев

    def merge_trees(self, t1, t2):
        """Сливает два дерева одинакового ранга (t1.key <= t2.key)"""
        if t1.key > t2.key:
            t1, t2 = t2, t1
        t2.parent = t1
        t1.children.append(t2)
        t1.degree += 1
        return t1

    def union(self, other):
        """Объединяет две биномиальные кучи"""
        # Сшиваем корни в один список, сортированный по degree
        combined = sorted(self.roots + other.roots, key=lambda x: x.degree)
        if not combined:
            return

        i = 0
        while i < len(combined) - 1:
            if combined[i].degree == combined[i + 1].degree:
                # Сливаем деревья одинакового ранга
                merged = self.merge_trees(combined[i], combined[i + 1])
                combined[i] = merged
                combined.pop(i + 1)
            else:
                i += 1
        self.roots = combined

    def insert(self, key):
        """Добавляет элемент"""
        new_heap = BinomialHeap()
        new_heap.roots.append(BinomialHeapNode(key))
        self.union(new_heap)

    def get_min(self):
        """Возвращает минимальный ключ (без удаления)"""
        if not self.roots:
            raise ValueError("Heap is empty")
        return min(node.key for node in self.roots)

    def extract_min(self):
        """Удаляет и возвращает минимальный элемент"""
        if not self.roots:
            raise ValueError("Heap is empty")

        # Находим дерево с минимальным корнем
        min_node = min(self.roots, key=lambda x: x.key)
        self.roots.remove(min_node)

        # Создаём новую кучу из детей min_node
        new_heap = BinomialHeap()
        for child in min_node.children:
            child.parent = None
            new_heap.roots.append(child)

        # Объединяем оставшуюся кучу с новой
        self.union(new_heap)
        return min_node.key

# Пример использования
bh = BinomialHeap()
bh.insert(10)
bh.insert(5)
bh.insert(20)
print(bh.get_min())      # 5
print(bh.extract_min())  # 5
print(bh.get_min())      # 10


 #**КУЧА ФИБОНАЧЧИ**
import math

class FibonacciHeapNode:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.mark = False
        # Двусвязный круговой список: left ↔ right
        self.left = self
        self.right = self


class FibonacciHeap:
    def __init__(self):
        self.min_node = None   # Указатель на узел с минимумом
        self.num_nodes = 0    # Число узлов в куче

    def is_empty(self):
        return self.min_node is None

    def insert(self, key):
        """Вставка за O(1)"""
        node = FibonacciHeapNode(key)
        if self.min_node is None:
            self.min_node = node
        else:
            # Вставляем node в корневой список (между min_node и min_node.right)
            node.left = self.min_node
            node.right = self.min_node.right
            self.min_node.right.left = node
            self.min_node.right = node
            # Обновляем min_node, если новый ключ меньше
            if node.key < self.min_node.key:
                self.min_node = node
        self.num_nodes += 1
        return node  # Возвращаем узел для decrease_key/delete


    def minimum(self):
        """Возврат минимума за O(1)"""
        if self.min_node is None:
            return None
        return self.min_node.key

    def merge(self, other_heap):
        """Слияние двух куч за O(1)"""
        if other_heap.is_empty():
            return
        if self.is_empty():
            self.min_node = other_heap.min_node
        else:
            # Сшиваем корневые списки
            a, b = self.min_node, other_heap.min_node
            a_right, b_left = a.right, b.left
            a.right = b
            b.left = a
            b_left.right = a_right
            a_right.left = b_left
            # Обновляем min_node
            if b.key < a.key:
                self.min_node = b
        self.num_nodes += other_heap.num_nodes

    def _link(self, parent, child):
        """Перемещает child в дети parent (используется в consolidate)"""
        # Удаляем child из корневого списка
        child.left.right = child.right
        child.right.left = child.left
        # Добавляем child к детям parent
        child.parent = parent
        if parent.child is None:
            parent.child = child
            child.left = child
            child.right = child
        else:
            child.left = parent.child
            child.right = parent.child.right
            parent.child.right.left = child
            parent.child.right = child
        parent.degree += 1
        child.mark = False

    def _consolidate(self):
        """Объединяет деревья одинаковой степени (вызывается из extract_min)"""
        if self.min_node is None:
            return
        # Максимальная степень дерева: log₂(n)
        max_degree = int(math.log(self.num_nodes) * 2) + 1
        degree_table = [None] * max_degree  # degree_table[d] — дерево степени d

        # Собираем все корневые узлы
        nodes = []
        current = self.min_node
        while True:
            nodes.append(current)
            current = current.right
            if current == self.min_node:
                break

        for node in nodes:
            d = node.degree
            while degree_table[d] is not None:
                other = degree_table[d]
                if node.key > other.key:
                    node, other = other, node  # node — меньший корень
                self._link(node, other)
                degree_table[d] = None
                d += 1
            degree_table[d] = node

        # Перестраиваем корневой список из degree_table
        self.min_node = None
        for tree in degree_table:
            if tree is not None:
                if self.min_node is None:
                    self.min_node = tree
                    tree.left = tree
                    tree.right = tree
                else:
                    # Вставляем tree в корневой список
                    tree.left = self.min_node
                    tree.right = self.min_node.right
                    self.min_node.right.left = tree
                    self.min_node.right = tree
                    if tree.key < self.min_node.key:
                        self.min_node = tree

    def extract_min(self):
        """Извлечение минимума за O(log n) (амортиз.)"""
        if self.min_node is None:
            raise ValueError("extract_min: куча пуста")

        z = self.min_node

        # 1. Добавляем детей z в корневой список
        if z.child is not None:
            child = z.child
            children = []
            while True:
                children.append(child)
                child = child.right
                if child == z.child:
                    break
            for child_node in children:
                # Отсоединяем от родителя и убираем mark
                child_node.parent = None
                child_node.mark = False
                # Вставляем в корневой список
                child_node.left = z
                child_node.right = z.right
                z.right.left = child_node
                z.left.right = child_node

        # 2. Удаляем z из корневого списка
        if z == z.right:
            self.min_node = None  # Куча стала пустой
        else:
            z.left.right = z.right
            z.right.left = z.left
            self.min_node = z.right  # Временный min — следующий
            self._consolidate()     # Объединяем деревья

        self.num_nodes -= 1
        return z.key

    def _cut(self, node, parent):
        """Отрезает node от parent и помещает в корневой список"""
        if node == node.right:
            parent.child = None
        else:
            node.left.right = node.right
            node.right.left = node.left
            if node == parent.child:
                parent.child = node.right
        parent.degree -= 1

        node.parent = None
        node.mark = False

        # Вставляем node в корневой список
        node.left = self.min_node
        node.right = self.min_node.right
        self.min_node.right.left = node
        self.min_node.right = node

    def _cascading_cut(self, node):
        """Каскадное отрезание вверх по родителям"""
        parent = node.parent
        if parent is not None:
            if not node.mark:
                node.mark = True
            else:
                self._cut(node, parent)
                self._cascading_cut(parent)

    def decrease_key(self, node, new_key):
        """Уменьшение ключа за O(1) (амортиз.)"""
        if new_key > node.key:
            raise ValueError("new_key > current key")
        node.key = new_key
        parent = node.parent

        if parent is not None and node.key < parent.key:
            self._cut(node, parent)
            self._cascading_cut(parent)

        if node.key < self.min_node.key:
            self.min




#**ХЭШ-ТАБЛИЦА**
class HashTable:
    def __init__(self, size=10):
        self.size = size
        self.table = [[] for _ in range(size)]  # Список списков для цепочек

    def _hash(self, key):
        """Простая хэш‑функция (сумма кодов символов по модулю размера)."""
        return sum(ord(c) for c in str(key)) % self.size

    def insert(self, key, value):
        """Добавить пару ключ‑значение."""
        index = self._hash(key)
        bucket = self.table[index]

        # Проверяем, есть ли ключ в цепочке
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)  # Обновляем значение
                return
        bucket.append((key, value))  # Добавляем новую пару

    def search(self, key):
        """Найти значение по ключу."""
        index = self._hash(key)
        bucket = self.table[index]

        for k, v in bucket:
            if k == key:
                return v
        raise KeyError(key)  # Ключ не найден

    def delete(self, key):
        """Удалить пару по ключу."""
        index = self._hash(key)
        bucket = self.table[index]

        for i, (k, v) in enumerate(bucket):
            if k == key:
                del bucket[i]
                return
        raise KeyError(key)  # Ключ не найден

    def __contains__(self, key):
        """Проверить наличие ключа (для оператора `in`)."""
        try:
            self.search(key)
            return True
        except KeyError:
            return False

    def __str__(self):
        """Строковое представление для отладки."""
        items = []
        for bucket in self.table:
            for k, v in bucket:
                items.append(f"{k}: {v}")
        return "{" + ", ".join(items) + "}"

# Пример использования
ht.insert("name", "Alice")
ht.insert("age", 25)
ht.insert("city", "Moscow")
...
print(ht)  # {name: Alice, age: 26}

File: лр 3/.cpp
   **БИНАРНАЯ КУЧА**
     
#include <iostream>
#include <vector>
#include <stdexcept>

class MaxHeap {
private:
    std::vector<int> heap;

    // Проталкивание элемента вверх (после вставки)
    void siftUp(int index) {
        if (index == 0) return;  // Корень — некуда подниматься

        int parent = (index - 1) / 2;
        if (heap[index] > heap[parent]) {
            std::swap(heap[index], heap[parent]);
            siftUp(parent);
        }
    }

    // Проталкивание элемента вниз (после извлечения)
    void siftDown(int index) {
        int left = 2 * index + 1;
        int right = 2 * index + 2;
        int largest = index;

        // Находим наибольший среди текущего, левого и правого потомка
        if (left < heap.size() && heap[left] > heap[largest])
            largest = left;
        if (right < heap.size() && heap[right] > heap[largest])
            largest = right;

        // Если наибольший не текущий — меняем и продолжаем
        if (largest != index) {
            std::swap(heap[index], heap[largest]);
            siftDown(largest);
        }
    }

public:
    // Вставка элемента
    void insert(int value) {
        heap.push_back(value);
        siftUp(heap.size() - 1);
    }

    // Извлечение максимального элемента
    int extractMax() {
        if (heap.empty())
            throw std::runtime_error("Heap is empty");

        int maxValue = heap[0];
        heap[0] = heap.back();  // Переносим последний элемент в корень
        heap.pop_back();      // Удаляем последний

        if (!heap.empty())     // Если куча не опустела — восстанавливаем свойство
            siftDown(0);

        return maxValue;
    }

    // Просмотр максимального элемента (без извлечения)
    int getMax() const {
        if (heap.empty())
            throw std::runtime_error("Heap is empty");
        return heap[0];
    }

    // Проверка на пустоту
    bool isEmpty() const {
        return heap.empty();
    }

    // Размер кучи
    size_t size() const {
        return heap.size();
    }

    // Вывод всех элементов (для отладки)
    void print() const {
        for (int val : heap) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
};

  Пример использования:
int main() {
    MaxHeap heap;

    // Вставляем элементы
    heap.insert(10);
    heap.insert(20);
    heap.insert(5);
    heap.insert(30);
    heap.insert(15);

    std::cout << "Куча: ";
    heap.print();  // 30 20 5 10 15

    std::cout << "Максимум: " << heap.getMax() << std::endl;  // 30

    std::cout << "Извлечён максимум: " << heap.extractMax() << std::endl;  // 30
    std::cout << "Новый максимум: " << heap.getMax() << std::endl;      // 20

    std::cout << "Куча после извлечения: ";
    heap.print();  // 20 15 5 10

    return 0;
}


**БИНОМИНАЛЬНАЯ КУЧА**
  
#include <iostream>
#include <vector>
#include <stdexcept>

// Узел биномиального дерева
template<typename T>
struct BinomialNode {
    T key;
    int degree;  // степень (количество детей)
    BinomialNode* parent;
    BinomialNode* child;   // первый ребёнок
    BinomialNode* sibling;  // следующий брат

    explicit BinomialNode(T k) : key(k), degree(0),
        parent(nullptr), child(nullptr), sibling(nullptr) {}
};

// Биномиальная куча
template<typename T>
class BinomialHeap {
private:
    BinomialNode<T>* head;  // голова списка корней

    // Слияние двух биномиальных деревьев одинаковой степени
    BinomialNode<T>* mergeTrees(BinomialNode<T>* t1, BinomialNode<T>* t2) {
        if (t1->key > t2->key) {
            std::swap(t1, t2);  // t1 получает меньший ключ
        }
        t2->parent = t1;
        t2->sibling = t1->child;
        t1->child = t2;
        t1->degree++;
        return t1;
    }

    // Консолидация кучи: объединение деревьев одинаковой степени
    void consolidate() {
        std::vector<BinomialNode<T>*> trees(64, nullptr);  // макс. степень ~64
        BinomialNode<T>* curr = head;
        head = nullptr;

        while (curr) {
            BinomialNode<T>* next = curr->sibling;
            curr->sibling = nullptr;  // отсоединяем от списка

            int d = curr->degree;
            while (trees[d]) {
                curr = mergeTrees(curr, trees[d]);
                trees[d] = nullptr;
                d++;
            }
            trees[d] = curr;
            curr = next;
        }

        // Собираем итоговый список корней
        for (auto& tree : trees) {
            if (tree) {
                tree->sibling = head;
                head = tree;
            }
        }
    }

public:
    BinomialHeap() : head(nullptr) {}

    // Вставка ключа (создаём кучу из 1 узла и сливаем)
    void insert(T key) {
        BinomialHeap temp;
        temp.head = new BinomialNode<T>(key);
        merge(temp);
    }

    // Объединение с другой кучей
    void merge(BinomialHeap& other) {
        // Сшиваем два списка корней в один (по возрастанию degree)
        BinomialNode<T>* newHead = nullptr;
        BinomialNode<T>** curr = &newHead;
        BinomialNode<T>* h1 = head;
        BinomialNode<T>* h2 = other.head;

        while (h1 && h2) {
            if (h1->degree <= h2->degree) {
                *curr = h1;
                h1 = h1->sibling;
            } else {
                *curr = h2;
                h2 = h2->sibling;
            }
            curr = &((*curr)->sibling);
        }
        *curr = h1 ? h1 : h2;

        head = newHead;
        other.head = nullptr;  // обнуляем вторую кучу
        consolidate();         // объединяем деревья одинаковой степени
    }

    // Извлечение минимального ключа
    T extractMin() {
        if (!head) throw std::runtime_error("Heap is empty");

        // Находим корень с минимальным ключом
        BinomialNode<T>* prevMin = nullptr;
        BinomialNode<T>* minNode = head;
        BinomialNode<T>* curr = head;

        while (curr->sibling) {
            if (curr->sibling->key < minNode->key) {
                prevMin = curr;
                minNode = curr->sibling;
            }
            curr = curr->sibling;
        }

        // Удаляем minNode из списка корней
        if (prevMin) {
            prevMin->sibling = minNode->sibling;
        } else {
            head = minNode->sibling;
        }

        // Создаём новую кучу из детей minNode
        BinomialHeap childrenHeap;
        BinomialNode<T>* child = minNode->child;
        while (child) {
            BinomialNode<T>* next = child->sibling;
            child->parent = nullptr;
            child->sibling = childrenHeap.head;
            childrenHeap.head = child;
            child = next;
        }

        T minKey = minNode->key;
        delete minNode;

        // Сливаем детей обратно в основную кучу
        merge(childrenHeap);
        return minKey;
    }

    // Уменьшение ключа узла (предполагаем, что узел доступен)
    void decreaseKey(BinomialNode<T>* node, T newKey) {
        if (newKey > node->key) {
            throw std::invalid_argument("New key is greater than current key");
        }
        node->key = newKey;

        // «Поднимаем» узел вверх, пока родитель больше
        BinomialNode<T>* parent = node->parent;
        while (parent && node->key < parent->key) {
            // Меняем ключи (не перестраиваем указатели!)
            std::swap(node->key, parent->key);
            node = parent;
            parent = node->parent;
        }
    }

    // Проверка на пустоту
    bool isEmpty() const {
        return head == nullptr;
    }

    // Деструктор (очистка памяти)
    ~BinomialHeap() {
        clear();
    }

    void clear() {
        while (!isEmpty()) {
            extractMin();
        }
    }
};

  Пример использования:
int main() {
    BinomialHeap<int> heap;

    // Вставка элементов
    heap.insert(10);
    heap.insert(5);
    heap.insert(20);
    heap.insert(3);
    heap.insert(15);

    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 3
    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 5

    heap.insert(1);
    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 1

    return 0;
}


  
**КУЧА ФИБОНАЧЧИ**
  
#include <iostream>
#include <vector>
#include <stdexcept>
#include <climits>

// Узел кучи Фибоначчи
template<typename T>
struct FibNode {
    T key;
    int degree;           // количество детей
    bool marked;          // отмечен ли узел (для каскадного вырезания)
    FibNode* parent;      // родитель
    FibNode* child;       // первый ребёнок
    FibNode* left;        // левый сосед в циклическом списке
    FibNode* right;       // правый сосед в циклическом списке

    explicit FibNode(T k) : key(k), degree(0), marked(false),
        parent(nullptr), child(nullptr), left(this), right(this) {}
};

// Куча Фибоначчи
template<typename T>
class FibonacciHeap {
private:
    FibNode<T>* min;     // указатель на узел с минимальным ключом
    int n;              // количество узлов в куче

    // Добавление узла в корневой список (циклический)
    void addToRootList(FibNode<T>* x) {
        if (!min) {
            min = x;
        } else {
            x->left = min;
            x->right = min->right;
            min->right->left = x;
            min->right = x;
            if (x->key < min->key) {
                min = x;
            }
        }
    }

    // Удаление узла из корневого списка
    void removeFromRootList(FibNode<T>* x) {
        x->left->right = x->right;
        x->right->left = x->left;
    }

    // Слияние корневых списков двух куч
    void mergeRootLists(FibNode<T>* h1, FibNode<T>* h2) {
        if (!h1 || !h2) return;
        FibNode<T>* temp = h1->right;
        h1->right = h2->right;
        h2->right->left = h1;
        h2->right = temp;
        temp->left = h2;
    }

    // Консолидация кучи (объединение деревьев одинаковой степени)
    void consolidate() {
        const int D = 64;  // макс. степень ~log(n)
        std::vector<FibNode<T>*> A(D, nullptr);

        std::vector<FibNode<T>*> roots;
        for (FibNode<T>* w = min; w; w = w->right) {
            roots.push_back(w);
            if (w->right == w) break;  // единственный узел
        }

        for (auto w : roots) {
            FibNode<T>* x = w;
            int d = x->degree;
            while (A[d]) {
                FibNode<T>* y = A[d];
                if (x->key > y->key) std::swap(x, y);
                link(y, x);
                A[d] = nullptr;
                d++;
            }
            A[d] = x;
        }

        min = nullptr;
        for (int i = 0; i < D; ++i) {
            if (A[i]) {
                addToRootList(A[i]);
            }
        }
    }

    // Присоединение дерева y к дереву x (y становится ребёнком x)
    void link(FibNode<T>* y, FibNode<T>* x) {
        removeFromRootList(y);
        y->parent = x;
        y->child = x->child;
        if (x->child) {
            y->child->left = y;
        }
        x->child = y;
        y->left = y->right = y;
        x->degree++;
        y->marked = false;
    }

    // Каскадное вырезание узла z
    void cascadingCut(FibNode<T>* z) {
        FibNode<T>* p = z->parent;
        if (p) {
            if (!z->marked) {
                z->marked = true;
            } else {
                cut(z, p);
                cascadingCut(p);
            }
        }
    }

    // Вырезание узла z из поддерева родителя p
    void cut(FibNode<T>* z, FibNode<T>* p) {
        // Удаляем z из списка детей p
        if (p->child == z) {
            p->child = z->right;
        }
        if (z->right != z) {
            z->left->right = z->right;
            z->right->left = z->left;
        }
        p->degree--;

        // Добавляем z в корневой список
        z->parent = nullptr;
        z->marked = false;
        addToRootList(z);
    }

public:
    FibonacciHeap() : min(nullptr), n(0) {}

    // Вставка ключа
    FibNode<T>* insert(T key) {
        FibNode<T>* x = new FibNode<T>(key);
        addToRootList(x);
        n++;
        return x;  // возвращаем узел для decreaseKey
    }

    // Извлечение минимального ключа
    T extractMin() {
        if (!min) throw std::runtime_error("Heap is empty");

        FibNode<T>* z = min;
        // Добавляем детей z в корневой список
        if (z->child) {
            FibNode<T>* child = z->child;
            do {
                child->parent = nullptr;
                child = child->right;
            } while (child != z->child);

            mergeRootLists(min, z->child);
        }

        // Удаляем z из корневого списка
        removeFromRootList(z);
        if (z == z->right) {  // последний узел
            min = nullptr;
        } else {
            min = z->right;
            consolidate();
        }

        T minKey = z->key;
        delete z;
        n--;
        return minKey;
    }

    // Уменьшение ключа узла x
    void decreaseKey(FibNode<T>* x, T k) {
        if (k > x->key) {
            throw std::invalid_argument("New key is greater than current key");
        }
        x->key = k;
        FibNode<T>* y = x->parent;

        if (y && x->key < y->key) {
            cut(x, y);
            cascadingCut(y);
        }
        if (x->key < min->key) {
            min = x;
        }
    }

    // Объединение с другой кучей
    void merge(FibonacciHeap& other) {
        mergeRootLists(min, other.min);
        if (!min || (other.min && other.min->key < min->key)) {
            min = other.min;
        }
        n += other.n;
        other.min = nullptr;
        other.n = 0;
    }

    // Проверка на пустоту
    bool isEmpty() const {
        return min == nullptr;
    }

    // Количество элементов
    int size() const {
        return n;
    }

    // Деструктор (очистка памяти)
    ~FibonacciHeap() {
        clear();
    }

    void clear() {
        while (!isEmpty()) {
            extractMin();
        }
    }
};


   Пример использования:
int main() {
    FibonacciHeap<int> heap;

    // Вставка элементов
    auto n1 = heap.insert(10);
    auto n2 = heap.insert(5);
    auto n3 = heap.insert(20);
    auto n4 = heap.insert(3);
    auto n5 = heap.insert(15);

    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 3
    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 5

    heap.decreaseKey(n3, 1);  // уменьшаем 20 до 1
    std::cout << "Minimum: " << heap.extractMin() << std::endl;  // 1

    return 0;
}

**ХЭШ-ТАБЛИЦЫ**
  
#include <iostream>
#include <vector>
#include <list>
#include <utility>

template<typename Key, typename Value>
class HashTable {
private:
    std::vector<std::list<std::pair<Key, Value>>> table;
    size_t capacity;
    size_t size;

    // Простая хеш‑функция (для строк и числовых типов)
    size_t hash(const Key& key) const {
        return std::hash<Key>{}(key) % capacity;
    }

    // Увеличение размера и перехеширование
    void rehash() {
        size_t newCapacity = capacity * 2;
        std::vector<std::list<std::pair<Key, Value>>> newTable(newCapacity);

        for (const auto& bucket : table) {
            for (const auto& pair : bucket) {
                size_t index = std::hash<Key>{}(pair.first) % newCapacity;
                newTable[index].push_back(pair);
            }
        }

        table = std::move(newTable);
        capacity = newCapacity;
    }

public:
    explicit HashTable(size_t initialCapacity = 8)
        : capacity(initialCapacity), size(0) {
        table.resize(capacity);
    }

    // Вставка пары (ключ, значение)
    void insert(const Key& key, const Value& value) {
        size_t index = hash(key);

        // Проверяем, есть ли ключ в цепочке
        for (auto& pair : table[index]) {
            if (pair.first == key) {
                pair.second = value;  // обновляем значение
                return;
            }
        }

        // Добавляем новый элемент
        table[index].push_back({key, value});
        size++;

        // Если коэффициент заполнения > 0.7, перехешируем
        if (size > capacity * 0.7) {
            rehash();
        }
    }

    // Поиск значения по ключу
    bool find(const Key& key, Value& value) const {
        size_t index = hash(key);
        for (const auto& pair : table[index]) {
            if (pair.first == key) {
                value = pair.second;
                return true;
            }
        }
        return false;
    }

    // Удаление по ключу
    bool remove(const Key& key) {
        size_t index = hash(key);
        auto& bucket = table[index];

        for (auto it = bucket.begin(); it != bucket.end(); ++it) {
            if (it->first == key) {
                bucket.erase(it);
                size--;
                return true;
            }
        }
        return false;
    }

    // Проверка наличия ключа
    bool contains(const Key& key) const {
        Value dummy;
        return find(key, dummy);
    }

    // Размер таблицы
    size_t getSize() const { return size; }
    size_t getCapacity() const { return capacity; }

    // Очистка
    void clear() {
        table.clear();
        table.resize(capacity);
        size = 0;
    }
};

  Пример использования:
int main() {
    HashTable<std::string, int> ht;

    // Вставка данных
    ht.insert("apple", 10);
    ht.insert("banana", 20);
    ht.insert("cherry", 30);

    // Поиск
    int value;
    if (ht.find("apple", value)) {
        std::cout << "apple: " << value << std::endl;
    }

    // Обновление
    ht.insert("apple", 15);  // перезапишет старое значение

    // Удаление
    if (ht.remove("banana")) {
        std::cout << "banana removed" << std::endl;
    }

    // Проверка наличия
    if (ht.contains("cherry")) {
        std::cout << "cherry exists" << std::endl;
    }

    std::cout << "Size: " << ht.getSize() << std::endl;

    return 0;
}
